### PR TITLE
Implement AVX/AVX2 BroadcastScalarToVector128/256

### PR DESCRIPTION
**Pull request by fiigii**
@CarolEidt @tannergooding

### CODE REVIEW

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx2.PlatformNotSupported.cs [hunk 1]
```diff
@@ -246,7 +246,7 @@ public static class Avx2
         /// __m128 _mm_broadcastss_ps (__m128 a)
         ///   VBROADCASTSS xmm, xmm
         /// __m128d _mm_broadcastsd_pd (__m128d a)
-        ///   VBROADCASTSD xmm, xmm
+        ///   VMOVDDUP xmm, xmm
```
Comment (fiigii) [inline src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx2.PlatformNotSupported.cs]: Fix a comment that AVX2 VBROADCASTSD cannot work with XMM registers.

File: tests/src/JIT/HardwareIntrinsics/X86/Shared/GenericUnOpTest.template [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+            );
+
+            Unsafe.Write(_dataTable.outArrayPtr, result);
+            ValidateResult(_dataTable.inArrayPtr, _dataTable.outArrayPtr);
+        }}
+
+        public void RunReflectionScenario_UnsafeRead()
+        {{
+            var result = typeof({0}).GetMethod(nameof({0}.{2}))
+                                     .MakeGenericMethod( new Type[] {{ typeof({6}) }})
```
Comment (fiigii) [inline tests/src/JIT/HardwareIntrinsics/X86/Shared/GenericUnOpTest.template]: Made a new template for generic intrinsics, which C# reflection has some limitations with generic methods.

**Comment by fiigii** [comment]
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

File: src/jit/hwintrinsiclistxarch.h [hunk 3]
```diff
@@ -372,6 +374,8 @@ HARDWARE_INTRINSIC(AVX2_And,                                         "And",
...
 HARDWARE_INTRINSIC(AVX2_BlendVariable,                               "BlendVariable",                                    AVX2,       -1,           32,           3,           {INS_vpblendvb, INS_vpblendvb, INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid},           HW_Category_SimpleSIMD,                        HW_Flag_NoFlag)
+HARDWARE_INTRINSIC(AVX2_BroadcastScalarToVector128,                  "BroadcastScalarToVector128",                       AVX2,       -1,           16,           1,           {INS_vpbroadcastb,INS_vpbroadcastb,INS_vpbroadcastw,INS_vpbroadcastw,INS_vpbroadcastd,INS_vpbroadcastd,INS_vpbroadcastq,INS_vpbroadcastq,INS_vbroadcastss,INS_movddup},        HW_Category_SimpleSIMD,         HW_Flag_OneTypeGeneric)
```
Comment (CarolEidt) [inline src/jit/hwintrinsiclistxarch.h]: At some point we'll want to clean up the formatting, but ideally as a formatting-only change.

File: tests/src/JIT/HardwareIntrinsics/X86/Avx/BroadcastScalarToVector128.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        const int Pass = 100;
+        const int Fail = 0;
+
+        static unsafe int Main(string[] args)
+        {
+            int testResult = Pass;
+
+            if (Avx.IsSupported)
+            {
+                using (TestTable<float> floatTable = new TestTable<float>(new float[8] { 1, -5, 100, 0, 1, 2, 3, 4 }, new float[4]))
```
Comment (tannergooding) [inline tests/src/JIT/HardwareIntrinsics/X86/Avx/BroadcastScalarToVector128.cs]: What was preventing this from being templated?

File: tests/src/JIT/HardwareIntrinsics/X86/Avx2/BroadcastScalarToVector128.Byte.cs [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+                Sse2.LoadAlignedVector128((Byte*)(_dataTable.inArrayPtr))
+            );
+
+            Unsafe.Write(_dataTable.outArrayPtr, result);
+            ValidateResult(_dataTable.inArrayPtr, _dataTable.outArrayPtr);
+        }
+
+        public void RunReflectionScenario_UnsafeRead()
+        {
+            var result = typeof(Avx2).GetMethod(nameof(Avx2.BroadcastScalarToVector128))
```
Comment (tannergooding) [inline tests/src/JIT/HardwareIntrinsics/X86/Avx2/BroadcastScalarToVector128.Byte.cs]: This won't work in all cases, just most.

There is at least one case where we have both generic and non-generic overloads, and this requires us to use `.GetMethods()` and then manually filter to the generic implementation.

Do you want to fix this now, or later?

File: src/jit/hwintrinsiclistxarch.h (hunk 3 above)
Comment (fiigii) [reply src/jit/hwintrinsiclistxarch.h]: Yes, I think we can make the formatting change after finish SSE2 and AVX.

File: tests/src/JIT/HardwareIntrinsics/X86/Avx/BroadcastScalarToVector128.cs (hunk 4 above)
Comment (fiigii) [reply tests/src/JIT/HardwareIntrinsics/X86/Avx/BroadcastScalarToVector128.cs]: Now, we do not have a template for `MemoryLoad` intrinsics. All the load/store intrinsics should be templated together later.

File: tests/src/JIT/HardwareIntrinsics/X86/Avx2/BroadcastScalarToVector128.Byte.cs (hunk 5 above)
Comment (fiigii) [reply tests/src/JIT/HardwareIntrinsics/X86/Avx2/BroadcastScalarToVector128.Byte.cs]: I just went over all the generic APIs, the only problem is `Test*` that are 2-op  intrinsics. So, this template (1-arg) is fine.

**Comment by fiigii** [comment]
@CarolEidt @tannergooding Thank you for the review, let's merge this PR.

**Pull request merged**